django/utils/encoding.py

```diff
@@ -196,20 +193,46 @@ def iri_to_uri(iri):
     return quote(force_bytes(iri), safe=b"/#%[]=:;$&()+,!?*@'~")
 
 
+_hexdig = '0123456789ABCDEFabcdef'
+_hextobyte = None
+
+
 def uri_to_iri(uri):
     """
     Converts a Uniform Resource Identifier(URI) into an Internationalized
     Resource Identifier(IRI).
 
-    This is the algorithm from section 3.2 of RFC 3987.
+    This is the algorithm from section 3.2 of RFC 3987, excluding step 4.
 
     Takes an URI in ASCII bytes (e.g. '/I%20%E2%99%A5%20Django/') and returns
     unicode containing the encoded result (e.g. '/I \xe2\x99\xa5 Django/').
     """
     if uri is None:
         return uri
     uri = force_bytes(uri)
-    iri = unquote_to_bytes(uri) if six.PY3 else unquote(uri)
+
+    bits = uri.split(b'%')
+    if len(bits) == 1:
+        iri = uri
+    else:
+        parts = [bits[0]]
+        append = parts.append
+        global _hextobyte
+        if _hextobyte is None:
+            _allowed_ascii = [45, 46, 95, 126] + list(range(65, 91)) + list(range(97, 123))
+            _hextobyte = {('%02x' % char).encode(): six.int2byte(char)
+                          for char in _allowed_ascii}
+            _hextobyte.update(((a + b).encode(), six.int2byte(int(a + b, 16)))
+                              for a in _hexdig[8:] for b in _hexdig)
```

Comment:
This code is based on the [`unquote` implementation in the standardlib](https://github.com/python/cpython/blob/436064bfe71ffd782dc5b942f65b80f7f0d7d2cd/Lib/urllib/parse.py#L554). I couldn't see any disadvantage of this approach, so I followed their example. But now that I'm looking at this again, this approach does not seem completely thread-safe.

The reason given in the standardlib is memory. I don't know the exact memory overhead for the strings and the dict, but I would assume this structure takes less than 4kb of memory.

If you think that's the better approach, I can move the initialization to module level.